src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,336 @@
+ *   - {@link #filteredNodes(Collection, Collection, Predicate)}: return the collection of FlowNodes matching the predicate.
+ *   - {@link #visitAll(Collection, FlowNodeVisitor)}: given a {@link FlowNodeVisitor}, invoke {@link FlowNodeVisitor#visit(FlowNode)} on each node and halt when it returns false.
+ *   - Iterator: Each FlowScanner can be used as an Iterator for FlowNode-by-FlowNode walking,
+ *               after you invoke {@link #setup(Collection, Collection)} to initialize it for iteration.
+ *   - {@link Filterator}: If initialized as an Iterator, each FlowScanner can provide a filtered view from the current point in time.
+ *   - Iterable: for syntactic sugar, FlowScanners implement Iterable to allow use in for-each loops once initialized.
+ *
+ * All APIs visit the parent nodes, walking backward from heads(inclusive) until they they hit {@link #blackList} nodes (exclusive) or reach the end of the DAG.
+ * If blackList nodes are an empty collection or null, APIs will walk to the beginning of the FlowGraph.
+ * Multiple blackList nodes are helpful for putting separate bounds on walking different parallel branches.
+ *
+ * Key Points:
+ *   - There are many helper methods offering syntactic sugar for the above APIs in common use cases (simpler method signatures).
+ *   - Each implementation provides its own iteration order (described in its javadoc comments).
+ *   - Implementations may visit some or all points in the DAG, this should be called out in the class's javadoc comments
+ *   - FlowScanners are NOT thread safe, for performance reasons and because it is too hard to guarantee.
+ *   - Many fields and methods are protected: this is intentional to allow building upon the implementations for more complex analyses.
+ *   - Each FlowScanner stores state internally for several reasons:
+ *      - This state can be used to construct more advanced analyses.
+ *      - FlowScanners can be reinitialized and reused repeatedly: the overheads of creating scanners repeatedly.
+ *
+ * Suggested uses:
+ *   - Implement a {@link FlowNodeVisitor} that collects metrics from each FlowNode visited, and call visitAll to extract the data.
+ *   - Find all flownodes of a given type (ex: stages), using {@link #filteredNodes(Collection, Collection, Predicate)}
+ *   - Find the first node with an Error before a specific node
+ *   - Scan through all nodes *just* within a block
+ *      - Use the {@link org.jenkinsci.plugins.workflow.graph.BlockEndNode} as the head
+ *      - Use the {@link org.jenkinsci.plugins.workflow.graph.BlockStartNode} as its blacklist with {@link Collections#singleton(Object)}
+ *
+ * TODO: come back and prettify this for HTML-style list formatting.
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
```

Comment:
:ant: cf. now-hidden discussion about the inadvisability of making an `Iterable` be its own `Iterator`